README.md
# LetterAvatarSimple

Generate letter image avatars based on user initials.

## Usage

```ruby
LetterAvatarSimple.generate("foobar")
# => #<StringIO:0x000055b3804948a8>
```

<img src="./examples/readme/F.png" alt='Image of letter "F"' width="64" height="64" />

------------

We can do multiple letters:

```ruby
LetterAvatarSimple.generate("Dallas Smith")
# => #<StringIO:0x000055b380e6a058>
```

<img src="./examples/readme/DS.png" alt='Image of letters "DS"' width="64" height="64" />

------------

We can save to a file:

```ruby
LetterAvatarSimple.generate_file("Dallas Smith")
# => #<File:/tmp/x20190527-19344-79m629.png>

LetterAvatarSimple.generate_file("Dallas Smith", filename: "/tmp/dallas.png")
# => #<File:/tmp/dallas.png>
```

------------

We can specify the initials using `LetterAvatarSimple::Identity`:

```ruby
i = LetterAvatarSimple::Identity.new("ZZ", "Dallas Smith")
LetterAvatarSimple.generate(i)
```

<img src="./examples/readme/ZZ-1.png" alt='Image of letters "ZZ" in green' width="64" height="64" />

------------

Image color is chosen from a palette based on username, so the same initials
will (most likely) not share the same color:

```ruby
i = LetterAvatarSimple::Identity.new("ZZ", "foobar")
LetterAvatarSimple.generate(i)
```

<img src="./examples/readme/ZZ-2.png" alt='Image of letters "ZZ" in pink' width="64" height="64" />

------------

The default palette is based on Google Inbox. We can change palettes:

```ruby
LetterAvatarSimple.generate("foobar", palette: :i_want_hue)
```

<img src="./examples/readme/F-i_want_hue.png" alt='Image of letter "F"' width="64" height="64" />

------------

Or skip the palette feature and provide the desired color directly, in RGB tuple format:

```ruby
LetterAvatarSimple.generate("foobar", color: [255,0,0])
```

<img src="./examples/readme/F-red.png" alt='Image of letter "F"' width="64" height="64" />

------------

Other options that can be provided to customize image generation:

```ruby
LetterAvatarSimple.generate(
  "foobar",
  size: 256,                            # => default 1024
  palette: :i_want_hue,                 # => default :google
  # warning: this bypasses the palette's color selection
  color: [255, 0, 0]                    # => default nil
  pointsize: 150,                       # => default 600
  font: "/tmp/path/to/font/file",       # => default is path to included Roboto font
  weight: 500,                          # => default 300
  fill_color: "rgba(255, 255, 255, 1)", # => default "rgba(255, 255, 255, 0.65)"
  annotate_position: "-0+10",           # => default "-0+5"
  filename: "/tmp/foo.png",             # => default is randomly generated tempfile path
)
```

## About

Forked from [letter_avatar][], which was in turn extracted from [Discourse][].

Compared to [letter_avatar][], this gem:
  * Outputs `StringIO` binary data by default (but can write to files too!)
    instead of always writing files to the `public` directory. This way you can
    use image upload gems like [shrine][] much easier
  * Supports keyword arguments for generating each image (don't need to edit
    global config or constants)
  * Simplifies custom palette loading and supports multiple custom palettes
  * Has some basic tests
  * Uses [minimagick][] instead of homegrown ImageMagick shell execution
  * Does **not** come with model, view, or controller helpers (you should
    be using [shrine][])
  * Does **not** do caching (you should be using [shrine][])

## Installation

```ruby
gem "letter_avatar_simple"
```

## System requirements

ImageMagick or GraphicsMagick - see
[MiniMagick requirements](https://github.com/minimagick/minimagick#requirements)

## Color palettes

Two color palettes are provided by default: `:google` and `:i_want_hue`

Both palettes use an MD5 digest of the username to select the color, so it's
likely that two different usernames that share the same initial(s) will render
with different colors.

If you need the same initials to always render with the same color, simply
provide a custom palette using a customized `letter_color` method. See below
for an example.

### `:google` - Google Inbox palette

<!--
alphabet = ("A".."Z").cycle

pairs = LetterAvatarSimple::Palettes::Google::PALETTE.zip(alphabet)

pairs.each do |color,letter|
  filename = "./examples/readme/google-palette/#{letter}.png"
  LetterAvatarSimple.generate_file(
    letter,
    color: color,
    filename: filename,
    size: 64,
    pointsize: 37.5,
    annotate_position: "-0+0"
  )
end
-->

<div>
<img src="./examples/readme/google-palette/A.png" alt='Image of letter "A"' width="64" height="64" />
<img src="./examples/readme/google-palette/B.png" alt='Image of letter "B"' width="64" height="64" />
<img src="./examples/readme/google-palette/C.png" alt='Image of letter "C"' width="64" height="64" />
<img src="./examples/readme/google-palette/D.png" alt='Image of letter "D"' width="64" height="64" />
<img src="./examples/readme/google-palette/E.png" alt='Image of letter "E"' width="64" height="64" />
<img src="./examples/readme/google-palette/F.png" alt='Image of letter "F"' width="64" height="64" />
<img src="./examples/readme/google-palette/G.png" alt='Image of letter "G"' width="64" height="64" />
<img src="./examples/readme/google-palette/H.png" alt='Image of letter "H"' width="64" height="64" />
<img src="./examples/readme/google-palette/I.png" alt='Image of letter "I"' width="64" height="64" />
<img src="./examples/readme/google-palette/J.png" alt='Image of letter "J"' width="64" height="64" />
<img src="./examples/readme/google-palette/K.png" alt='Image of letter "K"' width="64" height="64" />
<img src="./examples/readme/google-palette/L.png" alt='Image of letter "L"' width="64" height="64" />
<img src="./examples/readme/google-palette/M.png" alt='Image of letter "M"' width="64" height="64" />
<img src="./examples/readme/google-palette/N.png" alt='Image of letter "N"' width="64" height="64" />
<img src="./examples/readme/google-palette/O.png" alt='Image of letter "O"' width="64" height="64" />
<img src="./examples/readme/google-palette/P.png" alt='Image of letter "P"' width="64" height="64" />
<img src="./examples/readme/google-palette/Q.png" alt='Image of letter "Q"' width="64" height="64" />
<img src="./examples/readme/google-palette/R.png" alt='Image of letter "R"' width="64" height="64" />
<img src="./examples/readme/google-palette/S.png" alt='Image of letter "S"' width="64" height="64" />
<img src="./examples/readme/google-palette/T.png" alt='Image of letter "T"' width="64" height="64" />
<img src="./examples/readme/google-palette/U.png" alt='Image of letter "U"' width="64" height="64" />
<img src="./examples/readme/google-palette/V.png" alt='Image of letter "V"' width="64" height="64" />
<img src="./examples/readme/google-palette/W.png" alt='Image of letter "W"' width="64" height="64" />
<img src="./examples/readme/google-palette/X.png" alt='Image of letter "X"' width="64" height="64" />
<img src="./examples/readme/google-palette/Y.png" alt='Image of letter "Y"' width="64" height="64" />
<img src="./examples/readme/google-palette/Z.png" alt='Image of letter "Z"' width="64" height="64" />
</div>

### `:i_want_hue` - [iWantHue][] palette

<!--
alphabet = ("A".."Z").cycle

pairs = LetterAvatarSimple::Palettes::IWantHue::PALETTE.zip(alphabet)

f = File.open("/tmp/i_want_hue.txt", "w")
pairs.each_with_index do |(color,letter),i|
  filename = "./examples/readme/i_want_hue-palette/#{i}-#{letter}.png"
  LetterAvatarSimple.generate_file(
    letter,
    color: color,
    filename: filename,
    size: 64,
    pointsize: 37.5,
    annotate_position: "-0+0"
  )
  f.puts <<~HTML
    <img src="#{filename}" alt='Image of letter "#{letter}"' width="64" height="64" />
  HTML
end
f.close
-->

<div>
<img src="./examples/readme/i_want_hue-palette/0-A.png" alt='Image of letter "A"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/1-B.png" alt='Image of letter "B"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/2-C.png" alt='Image of letter "C"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/3-D.png" alt='Image of letter "D"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/4-E.png" alt='Image of letter "E"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/5-F.png" alt='Image of letter "F"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/6-G.png" alt='Image of letter "G"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/7-H.png" alt='Image of letter "H"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/8-I.png" alt='Image of letter "I"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/9-J.png" alt='Image of letter "J"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/10-K.png" alt='Image of letter "K"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/11-L.png" alt='Image of letter "L"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/12-M.png" alt='Image of letter "M"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/13-N.png" alt='Image of letter "N"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/14-O.png" alt='Image of letter "O"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/15-P.png" alt='Image of letter "P"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/16-Q.png" alt='Image of letter "Q"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/17-R.png" alt='Image of letter "R"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/18-S.png" alt='Image of letter "S"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/19-T.png" alt='Image of letter "T"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/20-U.png" alt='Image of letter "U"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/21-V.png" alt='Image of letter "V"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/22-W.png" alt='Image of letter "W"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/23-X.png" alt='Image of letter "X"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/24-Y.png" alt='Image of letter "Y"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/25-Z.png" alt='Image of letter "Z"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/26-A.png" alt='Image of letter "A"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/27-B.png" alt='Image of letter "B"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/28-C.png" alt='Image of letter "C"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/29-D.png" alt='Image of letter "D"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/30-E.png" alt='Image of letter "E"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/31-F.png" alt='Image of letter "F"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/32-G.png" alt='Image of letter "G"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/33-H.png" alt='Image of letter "H"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/34-I.png" alt='Image of letter "I"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/35-J.png" alt='Image of letter "J"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/36-K.png" alt='Image of letter "K"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/37-L.png" alt='Image of letter "L"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/38-M.png" alt='Image of letter "M"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/39-N.png" alt='Image of letter "N"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/40-O.png" alt='Image of letter "O"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/41-P.png" alt='Image of letter "P"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/42-Q.png" alt='Image of letter "Q"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/43-R.png" alt='Image of letter "R"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/44-S.png" alt='Image of letter "S"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/45-T.png" alt='Image of letter "T"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/46-U.png" alt='Image of letter "U"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/47-V.png" alt='Image of letter "V"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/48-W.png" alt='Image of letter "W"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/49-X.png" alt='Image of letter "X"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/50-Y.png" alt='Image of letter "Y"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/51-Z.png" alt='Image of letter "Z"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/52-A.png" alt='Image of letter "A"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/53-B.png" alt='Image of letter "B"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/54-C.png" alt='Image of letter "C"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/55-D.png" alt='Image of letter "D"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/56-E.png" alt='Image of letter "E"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/57-F.png" alt='Image of letter "F"' width="64" height="64" />
<img src="./examples/readme/i_want_hue-palette/58-G.png" alt='Image of letter "G"' width="64" height="64" />
... <a href="./examples/readme/i_want_hue-palette">and 157 more!</a>
</div>

### Custom palettes

You can add your own custom palette:

```ruby
LetterAvatarSimple.palettes[:my_palette] = LetterAvatarSimple::Palette.new([
  [120, 132, 205],
  [91, 149, 249],
  [72, 194, 249],
  [69, 208, 226],
])
# The default method of selecting the color is by MD5 digest of the username. You
# can change this behavior by providing a letter_color method.
LetterAvatarSimple.palettes[:my_palette].tap do |p|
  def p.letter_color(identity)
    if identity.id == "admin"
      [255, 0, 0] # red
    elsif identity.id == "bozo"
      @palette.sample # random
    else
      # same initials = same color
      digest = Digest::MD5.hexdigest(identity.letters.to_s)
      @palette[digest[0...15].to_i(16) % @palette.length]
    end
  end
end

LetterAvatarSimple.generate_file("foobar", palette: :my_palette)
```

## Configuration

The same options that can be passed to `generate` can be set as global defaults:

```ruby
LetterAvatarSimple.config do |config|
  config.size              = 256
  config.palette           = :i_want_hue
  config.color             = [255, 0, 0]
  config.pointsize         = 150
  config.font              = "/tmp/path/to/font/file"
  config.weight            = 500
  config.fill_color        = "rgba(255, 255, 255, 1)"
  config.annotate_position = "-0+10"
end
```

[letter_avatar]: https://github.com/ksz2k/letter_avatar
[minimagick]: https://github.com/minimagick/minimagick
[shrine]: https://github.com/shrinerb/shrine
[Discourse]: https://www.discourse.org/
[iWantHue]: http://tools.medialab.sciences-po.fr/iwanthue/index.php


CHANGELOG.md
1.0.0
-----

- Public-ready release
- Add color config option
- Change some defaults

0.9.1
-----

- Add keyword argument support to generate method
- Simplify custom palette specification
- Improve documentation

0.9.0
-----

- Initial release


spec/spec_helper.rb
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration

require 'bundler/setup'
Bundler.setup
Bundler.require

RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end

require "tempfile"

def stringio_to_file(s)
  Tempfile.new(['', '.png']).tap do |f|
    f.binmode
    f.write(s.read)
    f.close
  end
end

def image_fixture_path(image_path)
  File.expand_path("./fixtures/#{image_path}", File.dirname(__FILE__))
end


spec/letter_avatar_simple_spec.rb
require "spec_helper"

def image_compare(path1, path2)
  Imatcher::Matcher.new.compare(path1, path2)
end

describe "LetterAvatarSimple" do
  context "#generate" do
    context "with the default options" do
      context "given 'foobar'" do
        let(:image) { LetterAvatarSimple.generate("foobar") }
        let(:fixture_path) { image_fixture_path("foobar.png") }

        it "generates the correct image" do
          expect(image_compare(stringio_to_file(image).path, fixture_path)).to be_match
        end
      end

      context "given 'Dallas Smith'" do
        let(:image) { LetterAvatarSimple.generate("Dallas Smith") }
        let(:fixture_path) { image_fixture_path("Dallas Smith.png") }

        it "generates the correct image" do
          expect(image_compare(stringio_to_file(image).path, fixture_path)).to be_match
        end
      end
    end

    context "with a bunch of options provided" do
      let(:image) {
        LetterAvatarSimple.generate(
          "foobar",
          size: 256,
          palette: :i_want_hue,
          pointsize: 50,
          weight: 100,
          fill_color: "rgba(0, 255, 255, 1)",
          annotate_position: "-20+20",
        )
      }
      let(:fixture_path) { image_fixture_path("integration.png") }

      it "generates the correct image" do
        expect(image_compare(stringio_to_file(image).path, fixture_path)).to be_match
      end
    end
  end
end


lib/letter_avatar_simple.rb
# frozen_string_literal: true
require "letter_avatar_simple/configuration"
require "letter_avatar_simple/identity"
require "letter_avatar_simple/palette"
require "letter_avatar_simple/version"
require "fileutils"
require "mini_magick"
require "tmpdir"

class LetterAvatarSimple
  extend Configuration

  def self.config(&blk)
    yield(self)
  end

  def self.palettes
    @palettes
  end

  def self.palettes=(h)
    @palettes = h
  end

  def self.generate(identity, output: :string, **kwargs)
    if identity.is_a?(String)
      identity = Identity.from_username(identity)
    end

    opts = {
      palette: kwargs[:palette] || LetterAvatarSimple.palette,
      color: kwargs[:color] || LetterAvatarSimple.color,
      size: kwargs[:size] || LetterAvatarSimple.size,
      pointsize: kwargs[:pointsize] || LetterAvatarSimple.pointsize,
      font: kwargs[:font] || LetterAvatarSimple.font,
      weight: kwargs[:weight] || LetterAvatarSimple.weight,
      fill_color: kwargs[:fill_color] || LetterAvatarSimple.fill_color,
      annotate_position: kwargs[:annotate_position] || LetterAvatarSimple.annotate_position,
    }

    filename = kwargs[:filename] || generate_temp_filename
    color = opts[:color] || LetterAvatarSimple.palettes.fetch(opts.fetch(:palette)).letter_color(identity)
    MiniMagick::Tool::Convert.new do |x|
      x.size "#{opts.fetch(:size)}x#{opts.fetch(:size)}"
      x << "xc:#{to_rgb(color)}"
      x.pointsize opts.fetch(:pointsize)
      x.font opts.fetch(:font)
      x.weight opts.fetch(:weight)
      x.fill opts.fetch(:fill_color)
      x.gravity "Center"
      x.annotate(opts.fetch(:annotate_position), identity.letters)
      x << filename
    end

    if output == :file
      File.open(filename, "rb")
    else
      StringIO.new(File.binread(filename), "rb").tap do
        FileUtils.rm_f(filename)
      end
    end
  end

  def self.generate_file(*args, **kwargs)
    kwargs[:output] = :file
    self.generate(*args, **kwargs)
  end

  def self.to_rgb(color)
    r, g, b = color
    "rgb(#{r},#{g},#{b})"
  end

  def self.generate_temp_filename(ext=".png")
    filename = begin
      Dir::Tmpname.make_tmpname(["x", ext], nil)
    rescue NoMethodError
      require "securerandom"
      "#{SecureRandom.urlsafe_base64}#{ext}"
    end
    File.join(Dir.tmpdir, filename)
  end
end

require "letter_avatar_simple/palettes/google"
require "letter_avatar_simple/palettes/i_want_hue"
LetterAvatarSimple.palettes = {
  google: LetterAvatarSimple::Palettes::Google.new,
  i_want_hue: LetterAvatarSimple::Palettes::IWantHue.new,
}


lib/letter_avatar_simple/palettes/google.rb
# frozen_string_literal: true
class LetterAvatarSimple
  class Palettes
    class Google < LetterAvatarSimple::Palette
      # Colors from Google Inbox
      # https://inbox.google.com
      PALETTE = [
        [226, 95, 81], # A
        [242, 96, 145], # B
        [187, 101, 202], # C
        [149, 114, 207], # D
        [120, 132, 205], # E
        [91, 149, 249], # F
        [72, 194, 249], # G
        [69, 208, 226], # H
        [72, 182, 172], # I
        [82, 188, 137], # J
        [155, 206, 95], # K
        [212, 227, 74], # L
        [254, 218, 16], # M
        [247, 192, 0], # N
        [255, 168, 0], # O
        [255, 138, 96], # P
        [194, 194, 194], # Q
        [143, 164, 175], # R
        [162, 136, 126], # S
        [163, 163, 163], # T
        [175, 181, 226], # U
        [179, 155, 221], # V
        [194, 194, 194], # W
        [124, 222, 235], # X
        [188, 170, 164], # Y
        [173, 214, 125] # Z
      ].freeze

      def initialize
        @palette = PALETTE
      end
    end
  end
end


lib/letter_avatar_simple/colors.rb
# frozen_string_literal: true
class LetterAvatarSimple
  class Colors
    PALETTES = [:google, :iwanthue, :custom]

    GOOGLE_COLORS = [
      [226, 95, 81], # A
      [242, 96, 145], # B
      [187, 101, 202], # C
      [149, 114, 207], # D
      [120, 132, 205], # E
      [91, 149, 249], # F
      [72, 194, 249], # G
      [69, 208, 226], # H
      [72, 182, 172], # I
      [82, 188, 137], # J
      [155, 206, 95], # K
      [212, 227, 74], # L
      [254, 218, 16], # M
      [247, 192, 0], # N
      [255, 168, 0], # O
      [255, 138, 96], # P
      [194, 194, 194], # Q
      [143, 164, 175], # R
      [162, 136, 126], # S
      [163, 163, 163], # T
      [175, 181, 226], # U
      [179, 155, 221], # V
      [194, 194, 194], # W
      [124, 222, 235], # X
      [188, 170, 164], # Y
      [173, 214, 125] # Z
    ].freeze

    IWANTHUE_COLORS = [
      [198, 125, 40],
      [61, 155, 243],
      [74, 243, 75],
      [238, 89, 166],
      [52, 240, 224],
      [177, 156, 155],
      [240, 120, 145],
      [111, 154, 78],
      [237, 179, 245],
      [237, 101, 95],
      [89, 239, 155],
      [43, 254, 70],
      [163, 212, 245],
      [65, 152, 142],
      [165, 135, 246],
      [181, 166, 38],
      [187, 229, 206],
      [77, 164, 25],
      [179, 246, 101],
      [234, 93, 37],
      [225, 155, 115],
      [142, 140, 188],
      [223, 120, 140],
      [249, 174, 27],
      [244, 117, 225],
      [137, 141, 102],
      [75, 191, 146],
      [188, 239, 142],
      [164, 199, 145],
      [173, 120, 149],
      [59, 195, 89],
      [222, 198, 220],
      [68, 145, 187],
      [236, 204, 179],
      [159, 195, 72],
      [188, 121, 189],
      [166, 160, 85],
      [181, 233, 37],
      [236, 177, 85],
      [121, 147, 160],
      [234, 218, 110],
      [241, 157, 191],
      [62, 200, 234],
      [133, 243, 34],
      [88, 149, 110],
      [59, 228, 248],
      [183, 119, 118],
      [251, 195, 45],
      [113, 196, 122],
      [197, 115, 70],
      [80, 175, 187],
      [103, 231, 238],
      [240, 72, 133],
      [228, 149, 241],
      [180, 188, 159],
      [172, 132, 85],
      [180, 135, 251],
      [236, 194, 58],
      [217, 176, 109],
      [88, 244, 199],
      [186, 157, 239],
      [113, 230, 96],
      [206, 115, 165],
      [244, 178, 163],
      [230, 139, 26],
      [241, 125, 89],
      [83, 160, 66],
      [107, 190, 166],
      [197, 161, 210],
      [198, 203, 245],
      [238, 117, 19],
      [228, 119, 116],
      [131, 156, 41],
      [145, 178, 168],
      [139, 170, 220],
      [233, 95, 125],
      [87, 178, 230],
      [157, 200, 119],
      [237, 140, 76],
      [229, 185, 186],
      [144, 206, 212],
      [236, 209, 158],
      [185, 189, 79],
      [34, 208, 66],
      [84, 238, 129],
      [133, 140, 134],
      [67, 157, 94],
      [168, 179, 25],
      [140, 145, 240],
      [151, 241, 125],
      [67, 162, 107],
      [200, 156, 21],
      [169, 173, 189],
      [226, 116, 189],
      [133, 231, 191],
      [194, 161, 63],
      [241, 77, 99],
      [241, 217, 53],
      [123, 204, 105],
      [210, 201, 119],
      [229, 108, 155],
      [240, 91, 72],
      [187, 115, 210],
      [240, 163, 100],
      [178, 217, 57],
      [179, 135, 116],
      [204, 211, 24],
      [186, 135, 57],
      [223, 176, 135],
      [204, 148, 151],
      [116, 223, 50],
      [95, 195, 46],
      [123, 160, 236],
      [181, 172, 131],
      [142, 220, 202],
      [240, 140, 112],
      [172, 145, 164],
      [228, 124, 45],
      [135, 151, 243],
      [42, 205, 125],
      [192, 233, 116],
      [119, 170, 114],
      [158, 138, 26],
      [73, 190, 183],
      [185, 229, 243],
      [227, 107, 55],
      [196, 205, 202],
      [132, 143, 60],
      [233, 192, 237],
      [62, 150, 220],
      [205, 201, 141],
      [106, 140, 190],
      [161, 131, 205],
      [135, 134, 158],
      [198, 139, 81],
      [115, 171, 32],
      [101, 181, 67],
      [149, 137, 119],
      [37, 142, 183],
      [183, 130, 175],
      [168, 125, 133],
      [124, 142, 87],
      [236, 156, 171],
      [232, 194, 91],
      [219, 200, 69],
      [144, 219, 34],
      [219, 95, 187],
      [145, 154, 217],
      [165, 185, 100],
      [127, 238, 163],
      [224, 178, 198],
      [119, 153, 120],
      [124, 212, 92],
      [172, 161, 105],
      [231, 155, 135],
      [157, 132, 101],
      [122, 185, 146],
      [53, 166, 51],
      [70, 163, 90],
      [150, 190, 213],
      [210, 107, 60],
      [166, 152, 185],
      [159, 194, 159],
      [39, 141, 222],
      [202, 176, 161],
      [95, 140, 229],
      [168, 142, 87],
      [93, 170, 203],
      [159, 142, 54],
      [14, 168, 39],
      [94, 150, 149],
      [187, 206, 136],
      [157, 224, 166],
      [235, 158, 208],
      [109, 232, 216],
      [141, 201, 87],
      [208, 124, 118],
      [142, 125, 214],
      [19, 237, 174],
      [72, 219, 41],
      [234, 102, 111],
      [168, 142, 79],
      [188, 135, 35],
      [95, 155, 143],
      [148, 173, 116],
      [223, 112, 95],
      [228, 128, 236],
      [206, 114, 54],
      [195, 119, 88],
      [235, 140, 94],
      [235, 202, 125],
      [233, 155, 153],
      [214, 214, 238],
      [246, 200, 35],
      [151, 125, 171],
      [132, 145, 172],
      [131, 142, 118],
      [199, 126, 150],
      [61, 162, 123],
      [58, 176, 151],
      [215, 141, 69],
      [225, 154, 220],
      [220, 77, 167],
      [233, 161, 64],
      [130, 221, 137],
      [81, 191, 129],
      [169, 162, 140],
      [174, 177, 222],
      [236, 174, 47],
      [233, 188, 180],
      [69, 222, 172],
      [71, 232, 93],
      [118, 211, 238],
      [157, 224, 83],
      [218, 105, 73],
      [126, 169, 36]
    ].freeze

    def self.for(username)
      public_send("with_#{LetterAvatarSimple.colors_palette}", username)
    end

    def self.with_iwanthue(username)
      iwanthue[
        Digest::MD5.hexdigest(username)[0...15].to_i(16) % iwanthue.length
      ]
    end

    def self.with_google(username)
      char = username[0].upcase

      if /[A-Z]/ =~ char
        # 65 is 'A' ord
        idx = char.ord - 65
        google[idx]
      elsif /[\d]/ =~ char
        google[char.to_i]
      else
        google[Digest::MD5.hexdigest(username)[0...15].to_i(16) % google.length]
      end
    end

    def self.with_custom(username)
      custom_palette = LetterAvatarSimple.custom_palette
      custom_palette[Digest::MD5.hexdigest(username)[0...15].to_i(16) % custom_palette.length]
    end
    
    def self.valid_custom_palette?(palette)
      return false unless palette.is_a?(Array)
      return palette.all? do |color| 
        false unless color.is_a?(Array)
        color.all? { |i| i.is_a?(Integer) }
      end
    end

    # Colors form Google Inbox
    # https://inbox.google.com
    def self.google
      GOOGLE_COLORS
    end

    # palette of optimally disctinct colors
    # cf. http://tools.medialab.sciences-po.fr/iwanthue/index.php
    # parameters used:
    #   - H: 0 - 360
    #   - C: 0 - 2
    #   - L: 0.75 - 1.5
    def self.iwanthue
      IWANTHUE_COLORS
    end
  end
end
